ping/rust/Dockerfile

```diff
@@ -4,43 +4,23 @@ WORKDIR /usr/src/testplan
 RUN apt-get update && apt-get install -y cmake protobuf-compiler
```

Comment:
@laurentsenta tagging you here for visibility, given that you have worked on the caching optimizations below.

With this pull request `cargo build` is the same across rust-libp2p versions (apart from `master` and a specific git commit hash), i.e. it builds all rust-libp2p versions and can thus be reused when testing different rust-libp2p versions. This results in the below simplifications.

Let us know in case you have any objections.

Reply:
@mxinden thanks for the heads up; no objection on my side, but a question and a comment:

(q) it seems like we "just" moved the version indirection from feature flags to one layer up, but no big change in the `Cargo.toml`:

So we replace `cargo build --features="${VERSION_ID}"` with `cargo build --bin=${VERSION_ID}`

Why is this improving how we track dependencies in the `Cargo.lock`?

(c) We're dropping dependency caching, i.e. "do not download and rebuild all my dependencies when I change one line in my main.rs".

I added this to make the experience of iterating on the test code more bearable. We might want to add back that feature eventually (through dummy mains, `cargo-chef`, or something else).

Reply:
Got it; thanks for the clarification and for updating the description

> If somehow possible, I'd like to avoid containers for local iteration. Can testground run local binaries too?

Yes, it doesn't do network isolation and simulation, so you might not be able to use this as the main workflow to develop & fix tests.

> What is important is splitting the binaries per version to allow fine granular control over the code per version.

One last comment on this, if that's the point of this PR, is it possible we're trading "simple" for "easy" and we'll pay in the long term?

We're replacing "I have one main, with many shims for backward compatibility" with "I have many mains re-implementing a shared API".

The go version has been around for longer, so we have more examples of this:

- [One shared main](https://github.com/libp2p/test-plans/blob/b97b36aa2754c77795635facbf96c8504bdd9135/ping/go/main.go)
- [Shim some parts of the library](https://github.com/libp2p/test-plans/blob/b97b36aa2754c77795635facbf96c8504bdd9135/ping/go/compat/libp2p.v0.20.go#L1-L2)  (build flags in the header are ~= rust feature flags)

Note how surgical we can be when we support new versions, upgrading from v0.20 to v0.21 is "just" adding a flag because there was no API-breaking change.

With the new approach, it seems like we have 7 main files, ~600 lines, and only ~40 lines that are different.

Reply:
> One last comment on this, if that's the point of this PR, is it possible we're trading "simple" for "easy" and we'll pay in the long term?

I see your point. Though unfortunately I don't see how we can support future breaking changes in rust-libp2p with the current "simple" approach. In other words, I don't see an alternative to the "easy" approach going forward.

All that said, I think it is relatively easy to iterate on this part of the code base. In case we come up with a better design, let's refactor `ping/rust`.

> > (c) We're dropping dependency caching, i.e. "do not download and rebuild all my dependencies when I change one line in my main.rs".
> > I added this to make the experience of iterating on the test code more bearable. We might want to add back that feature eventually (through dummy mains, `cargo-chef`, or something else).
> 
> Yeah I've been there. I am okay with it being added back. I stripped it to simplify things as much as possible during debugging.

@thomaseizinger would you mind adding it back in? I suggest not blocking this pull request, but instead do so in a follow-up pull request.

Reply:
> > Yeah I've been there. I am okay with it being added back. I stripped it to simplify things as much as possible during debugging.
> 
> @thomaseizinger would you mind adding it back in? I suggest not blocking this pull request, but instead do so in a follow-up pull request.

I'll have a look. Right now, I am just happy to be unblocked on my PR :)